File: interface-graphique/analysis.ts
import * as fs from 'fs';
import * as path from 'path';

// Configuration interfaces
interface AnalysisConfig {
  enabled: boolean;
  mlEnabled: boolean;
  sensitivity: number;
  timeWindow: number;
  minDataPoints: number;
  maxDataPoints: number;
  updateInterval: number;
  batchSize: number;
  confidenceThreshold: number;
}

interface ThresholdsConfig {
  normal: number;
  warning: number;
  critical: number;
  emergency: number;
  baseline: number;
  spikeThreshold: number;
  trendThreshold: number;
}

interface MLModelConfig {
  type: string;
  contamination: number;
  nEstimators: number;
  maxSamples: string;
  randomState: number;
  nJobs: number;
  maxFeatures: number;
  bootstrap: boolean;
  warmStart: boolean;
  verbose: number;
}

// Traffic data interface
interface TrafficData {
  timestamp: Date;
  value: number;
  sourceIp: string;
  destinationIp: string;
  protocol: string;
  port: number;
  bytesSent: number;
  bytesReceived: number;
  packetsSent: number;
  packetsReceived: number;
}

// Features interface
interface Features {
  mean: number;
  std: number;
  max: number;
  min: number;
  median: number;
  q25: number;
  q75: number;
  rateOfChange: number;
}

// Pattern interface
interface Pattern {
  type: string;
  count: number;
  maxValue: number;
  timestamps: Date[];
  slope: number;
}

// Threat level enum
enum ThreatLevel {
  NORMAL = 'NORMAL',
  NOTICE = 'NOTICE',
  WARNING = 'WARNING',
  CRITICAL = 'CRITICAL',
  EMERGENCY = 'EMERGENCY'
}

// ML model interface
interface MLModel {
  contaminationScores: number[];
  sensitivity: number;
  trained: boolean;
}

// Traffic Analyzer class
export class TrafficAnalyzer {
  private analysisConfig: AnalysisConfig;
  private thresholdsConfig: ThresholdsConfig;
  private mlModelConfig: MLModelConfig;
  private trafficHistory: TrafficData[] = [];
  private modelData: number[] = [];
  private scalerData: number[] = [];
  private mlModel: MLModel;
  private logWriter: fs.WriteStream | null = null;
  private rulesDir: string;
  private modelLoaded: boolean = false;

  constructor() {
    this.analysisConfig = {
      enabled: true,
      mlEnabled: true,
      sensitivity: 0.1,
      timeWindow: 300,
      minDataPoints: 10,
      maxDataPoints: 10000,
      updateInterval: 60,
      batchSize: 100,
      confidenceThreshold: 0.8
    };

    this.thresholdsConfig = {
      normal: 50,
      warning: 100,
      critical: 200,
      emergency: 500,
      baseline: 30,
      spikeThreshold: 2.0,
      trendThreshold: 0.1
    };

    this.mlModelConfig = {
      type: 'isolation_forest',
      contamination: 0.1,
      nEstimators: 100,
      maxSamples: 'auto',
      randomState: 42,
      nJobs: -1,
      maxFeatures: 1.0,
      bootstrap: false,
      warmStart: false,
      verbose: 0
    };

    this.mlModel = {
      contaminationScores: new Array(1000).fill(0),
      sensitivity: this.analysisConfig.sensitivity,
      trained: false
    };

    this.initialize();
  }

  private initialize(): void {
    // Create rules directory
    this.rulesDir = 'rules';
    if (!fs.existsSync(this.rulesDir)) {
      fs.mkdirSync(this.rulesDir, { recursive: true });
    }

    // Initialize logging
    const logPath = path.join(this.rulesDir, 'traffic_analyzer.log');
    this.logWriter = fs.createWriteStream(logPath, { flags: 'a' });

    // Load or create ML model
    this.loadModel();

    // Initialize ML model
    this.mlModel.sensitivity = this.analysisConfig.sensitivity;

    this.log('TrafficAnalyzer initialized successfully');
  }

  private log(message: string): void {
    const timestamp = new Date().toISOString();
    const logMessage = `[${timestamp}] ${message}\n`;

    if (this.logWriter) {
      this.logWriter.write(logMessage);
    }

    console.log(logMessage.trim());
  }

  private loadModel(): void {
    const modelPath = path.join(this.rulesDir, 'traffic_model.bin');

    if (fs.existsSync(modelPath)) {
      try {
        const modelBytes = fs.readFileSync(modelPath);
        this.modelData = [];

        // Read model data (simplified binary format)
        for (let i = 0; i < modelBytes.length; i += 8) {
          if (i + 8 <= modelBytes.length) {
            const value = modelBytes.readDoubleLE(i);
            this.modelData.push(value);
          }
        }

        this.modelLoaded = true;
        this.log('ML model loaded successfully');
      } catch (error) {
        this.log(`Error loading ML model: ${error}`);
        this.createNewModel();
      }
    } else {
      this.createNewModel();
    }
  }

  private createNewModel(): void {
    this.modelData = [];
    for (let i = 0; i < 1000; i++) {
      this.modelData.push(Math.random());
    }

    this.modelLoaded = true;
    this.log('New ML model created');
  }

  private saveModel(): void {
    try {
      const modelPath = path.join(this.rulesDir, 'traffic_model.bin');
      const buffer = Buffer.alloc(this.modelData.length * 8);

      for (let i = 0; i < this.modelData.length; i++) {
        buffer.writeDoubleLE(this.modelData[i], i * 8);
      }

      fs.writeFileSync(modelPath, buffer);
      this.log('ML model saved successfully');
    } catch (error) {
      this.log(`Error saving ML model: ${error}`);
    }
  }

  public extractFeatures(trafficData: number[]): Features {
    const features: Features = {
      mean: 0,
      std: 0,
      max: 0,
      min: 0,
      median: 0,
      q25: 0,
      q75: 0,
      rateOfChange: 0
    };

    if (trafficData.length === 0) {
      return features;
    }

    // Calculate basic statistics
    const sum = trafficData.reduce((acc, val) => acc + val, 0);
    features.mean = sum / trafficData.length;
    features.min = Math.min(...trafficData);
    features.max = Math.max(...trafficData);

    // Calculate standard deviation
    const sumSq = trafficData.reduce((acc, val) => acc + val * val, 0);
    const variance = (sumSq / trafficData.length) - (features.mean * features.mean);
    features.std = Math.sqrt(Math.max(0, variance));

    // Calculate percentiles
    const sortedData = [...trafficData].sort((a, b) => a - b);
    features.median = sortedData[Math.floor(sortedData.length / 2)];
    features.q25 = sortedData[Math.floor(sortedData.length / 4)];
    features.q75 = sortedData[Math.floor(3 * sortedData.length / 4)];

    // Calculate rate of change
    if (trafficData.length > 1) {
      let totalChange = 0;
      for (let i = 1; i < trafficData.length; i++) {
        totalChange += trafficData[i] - trafficData[i - 1];
      }
      features.rateOfChange = totalChange / (trafficData.length - 1);
    }

    return features;
  }

  public analyzeTraffic(currentTraffic: number): { isAnomaly: boolean; score: number } {
    // Add current traffic to history
    const data: TrafficData = {
      timestamp: new Date(),
      value: currentTraffic,
      sourceIp: '',
      destinationIp: '',
      protocol: '',
      port: 0,
      bytesSent: 0,
      bytesReceived: 0,
      packetsSent: 0,
      packetsReceived: 0
    };

    this.trafficHistory.push(data);

    // Clean old data
    const cutoffTime = new Date(Date.now() - this.analysisConfig.timeWindow * 1000);
    this.trafficHistory = this.trafficHistory.filter(entry => entry.timestamp > cutoffTime);

    // Extract traffic values
    const trafficValues = this.trafficHistory.map(entry => entry.value);

    // Extract features
    const features = this.extractFeatures(trafficValues);

    // Predict anomaly
    if (this.analysisConfig.mlEnabled) {
      return this.predictAnomalyML(features);
    } else {
      return this.predictAnomalyBasic(features);
    }
  }

  private predictAnomalyML(features: Features): { isAnomaly: boolean; score: number } {
    // Simulate ML prediction
    const score = this.calculateAnomalyScore(features);
    const isAnomaly = score < -this.analysisConfig.sensitivity;

    return { isAnomaly, score: Math.abs(score) };
  }

  private predictAnomalyBasic(features: Features): { isAnomaly: boolean; score: number } {
    const isAnomaly = features.mean > this.thresholdsConfig.warning ||
                     features.max > this.thresholdsConfig.critical;

    const score = features.mean / this.thresholdsConfig.warning;

    return { isAnomaly, score };
  }

  private calculateAnomalyScore(features: Features): number {
    // Simulate isolation forest scoring
    let score = 0.0;

    // Combine features with weights
    score += features.mean * 0.3;
    score += features.std * 0.2;
    score += features.max * 0.2;
    score += features.rateOfChange * 0.3;

    // Normalize score
    score = (score - 50.0) / 25.0; // Assuming normal range

    return -score; // Negative for anomalies
  }

  public detectPatterns(trafficHistory: TrafficData[]): Pattern[] {
    const patterns: Pattern[] = [];

    if (trafficHistory.length < 10) {
      return patterns;
    }

    // Extract values
    const values = trafficHistory.map(entry => entry.value);
    const timestamps = trafficHistory.map(entry => entry.timestamp);

    // Detect spikes
    this.detectSpikes(values, timestamps, patterns);

    // Detect trends
    this.detectTrends(values, timestamps, patterns);

    return patterns;
  }

  private detectSpikes(values: number[], timestamps: Date[], patterns: Pattern[]): void {
    if (values.length < 5) return;

    // Calculate rolling mean and std
    const rollingMean: number[] = [];
    const rollingStd: number[] = [];
    const window = 5;

    for (let i = window; i < values.length; i++) {
      const windowValues = values.slice(i - window, i);
      const mean = windowValues.reduce((acc, val) => acc + val, 0) / window;
      const variance = windowValues.reduce((acc, val) => acc + Math.pow(val - mean, 2), 0) / window;
      const std = Math.sqrt(Math.max(0, variance));

      rollingMean.push(mean);
      rollingStd.push(std);
    }

    // Detect anomalies
    for (let i = 0; i < rollingMean.length; i++) {
      const diff = Math.abs(values[i + window] - rollingMean[i]);
      if (diff > 2 * rollingStd[i]) {
        const pattern: Pattern = {
          type: 'spike',
          count: 1,
          maxValue: values[i + window],
          timestamps: [timestamps[i + window]],
          slope: 0
        };
        patterns.push(pattern);
      }
    }
  }

  private detectTrends(values: number[], timestamps: Date[], patterns: Pattern[]): void {
    if (values.length < 30) return;

    // Calculate trend using linear regression
    let sumX = 0.0, sumY = 0.0, sumXY = 0.0, sumX2 = 0.0;
    const n = values.length;

    for (let i = 0; i < n; i++) {
      sumX += i;
      sumY += values[i];
      sumXY += i * values[i];
      sumX2 += i * i;
    }

    const slope = (n * sumXY - sumX * sumY) / (n * sumX2 - sumX * sumX);

    if (Math.abs(slope) > this.thresholdsConfig.trendThreshold) {
      const pattern: Pattern = {
        type: 'trend',
        count: 0,
        maxValue: 0,
        timestamps: [],
        slope
      };
      patterns.push(pattern);
    }
  }

  public saveAnalysis(analysisData: any): string {
    try {
      const filename = `analysis_${new Date().toISOString().replace(/[:.]/g, '-')}.json`;
      const filepath = path.join(this.rulesDir, filename);

      fs.writeFileSync(filepath, JSON.stringify(analysisData, null, 2));

      this.log(`Analysis saved to ${filepath}`);
      return filepath;
    } catch (error) {
      this.log(`Error saving analysis: ${error}`);
      return '';
    }
  }

  public getThreatLevel(anomalyScore: number): ThreatLevel {
    if (anomalyScore > 0.8) return ThreatLevel.CRITICAL;
    if (anomalyScore > 0.6) return ThreatLevel.WARNING;
    if (anomalyScore > 0.4) return ThreatLevel.NOTICE;
    return ThreatLevel.NORMAL;
  }

  public updateModel(newData: TrafficData[]): boolean {
    if (newData.length < 100) {
      return false;
    }

    try {
      const values = newData.map(entry => entry.value);

      // Update model data (simplified)
      for (let i = 0; i < Math.min(values.length, this.modelData.length); i++) {
        this.modelData[i] = (this.modelData[i] + values[i]) / 2.0;
      }

      this.saveModel();
      this.log('ML model updated successfully');
      return true;
    } catch (error) {
      this.log(`Error updating model: ${error}`);
      return false;
    }
  }

  // Configuration methods
  public setSensitivity(sensitivity: number): void {
    this.analysisConfig.sensitivity = sensitivity;
    this.mlModel.sensitivity = sensitivity;
  }

  public getSensitivity(): number {
    return this.analysisConfig.sensitivity;
  }

  public setMlEnabled(enabled: boolean): void {
    this.analysisConfig.mlEnabled = enabled;
  }

  public isMlEnabled(): boolean {
    return this.analysisConfig.mlEnabled;
  }

  // Statistics methods
  public getHistorySize(): number {
    return this.trafficHistory.length;
  }

  public getRecentTraffic(count: number = 100): TrafficData[] {
    const start = Math.max(0, this.trafficHistory.length - count);
    return this.trafficHistory.slice(start);
  }

  public dispose(): void {
    if (this.logWriter) {
      this.logWriter.end();
    }
  }
}

// Main function for testing
export async function main(): Promise<void> {
  const analyzer = new TrafficAnalyzer();

  // Simulate traffic data
  for (let i = 0; i < 100; i++) {
    let trafficValue: number;

    // Simulate occasional anomalies
    if (i % 20 === 0) {
      trafficValue = Math.random() * 100 + 100; // 100-200
    } else {
      trafficValue = Math.random() * 20 + 40; // 40-60
    }

    const { isAnomaly, score } = analyzer.analyzeTraffic(trafficValue);

    if (isAnomaly) {
      console.log(`ANOMALY DETECTED! Score: ${score}`);
    }

    await new Promise(resolve => setTimeout(resolve, 100));
  }

  analyzer.dispose();
}

// Run if this file is executed directly
if (require.main === module) {
  main().catch(console.error);
} 

File: backend/cpp/cpp/analysis.cpp
#include <iostream>
#include <vector>
#include <map>
#include <chrono>
#include <thread>
#include <fstream>
#include <sstream>
#include <algorithm>
#include <cmath>
#include <random>
#include <memory>
#include <mutex>
#include <condition_variable>
#include <queue>
#include <functional>
#include <filesystem>
#include <json/json.h>

namespace fs = std::filesystem;

// Configuration structure
struct Config {
    struct Analysis {
        bool enabled = true;
        bool ml_enabled = true;
        double sensitivity = 0.1;
        int time_window = 300;
        int min_data_points = 10;
        int max_data_points = 10000;
        int update_interval = 60;
        int batch_size = 100;
        double confidence_threshold = 0.8;
    } analysis;

    struct Thresholds {
        int normal = 50;
        int warning = 100;
        int critical = 200;
        int emergency = 500;
        int baseline = 30;
        double spike_threshold = 2.0;
        double trend_threshold = 0.1;
    } thresholds;

    struct MLModel {
        std::string type = "isolation_forest";
        double contamination = 0.1;
        int n_estimators = 100;
        std::string max_samples = "auto";
        int random_state = 42;
        int n_jobs = -1;
        double max_features = 1.0;
        bool bootstrap = false;
        bool warm_start = false;
        int verbose = 0;
    } ml_model;
};

// Traffic data structure
struct TrafficData {
    std::chrono::system_clock::time_point timestamp;
    double value;
    std::string source_ip;
    std::string destination_ip;
    std::string protocol;
    int port;
    size_t bytes_sent;
    size_t bytes_received;
    size_t packets_sent;
    size_t packets_received;
};

// Features structure
struct Features {
    double mean = 0.0;
    double std = 0.0;
    double max = 0.0;
    double min = 0.0;
    double median = 0.0;
    double q25 = 0.0;
    double q75 = 0.0;
    double rate_of_change = 0.0;
};

// Pattern structure
struct Pattern {
    std::string type;
    int count = 0;
    double max_value = 0.0;
    std::vector<std::chrono::system_clock::time_point> timestamps;
    double slope = 0.0;
};

// Threat level enum
enum class ThreatLevel {
    NORMAL,
    NOTICE,
    WARNING,
    CRITICAL,
    EMERGENCY
};

// Traffic Analyzer class
class TrafficAnalyzer {
private:
    Config config;
    std::vector<TrafficData> traffic_history;
    std::mutex history_mutex;
    std::vector<double> model_data;
    std::vector<double> scaler_data;
    std::mt19937 rng;
    std::ofstream log_file;
    std::string rules_dir;
    bool model_loaded = false;

    // ML model simulation
    struct MLModel {
        std::vector<double> contamination_scores;
        double sensitivity;
        bool trained = false;
    } ml_model;

public:
    TrafficAnalyzer() : rng(std::random_device{}()) {
        initialize();
    }

    ~TrafficAnalyzer() {
        if (log_file.is_open()) {
            log_file.close();
        }
    }

private:
    void initialize() {
        // Create rules directory
        rules_dir = "rules";
        fs::create_directories(rules_dir);

        // Initialize logging
        std::string log_path = rules_dir + "/traffic_analyzer.log";
        log_file.open(log_path, std::ios::app);
        
        // Load or create ML model
        load_model();
        
        // Initialize ML model
        ml_model.sensitivity = config.analysis.sensitivity;
        ml_model.contamination_scores.resize(1000, 0.0);
        
        log("TrafficAnalyzer initialized successfully");
    }

    void log(const std::string& message) {
        auto now = std::chrono::system_clock::now();
        auto time_t = std::chrono::system_clock::to_time_t(now);
        std::string timestamp = std::ctime(&time_t);
        timestamp.pop_back(); // Remove newline
        
        std::string log_message = "[" + timestamp + "] " + message + "\n";
        
        if (log_file.is_open()) {
            log_file << log_message;
            log_file.flush();
        }
        
        std::cout << log_message;
    }

    void load_model() {
        std::string model_path = rules_dir + "/traffic_model.bin";
        
        if (fs::exists(model_path)) {
            try {
                std::ifstream file(model_path, std::ios::binary);
                if (file.is_open()) {
                    // Read model data
                    size_t size;
                    file.read(reinterpret_cast<char*>(&size), sizeof(size));
                    model_data.resize(size);
                    file.read(reinterpret_cast<char*>(model_data.data()), size * sizeof(double));
                    
                    model_loaded = true;
                    log("ML model loaded successfully");
                }
            } catch (const std::exception& e) {
                log("Error loading ML model: " + std::string(e.what()));
                create_new_model();
            }
        } else {
            create_new_model();
        }
    }

    void create_new_model() {
        model_data.clear();
        model_data.resize(1000, 0.0);
        
        // Initialize with random data
        std::uniform_real_distribution<double> dist(0.0, 1.0);
        for (auto& value : model_data) {
            value = dist(rng);
        }
        
        model_loaded = true;
        log("New ML model created");
    }

    void save_model() {
        try {
            std::string model_path = rules_dir + "/traffic_model.bin";
            std::ofstream file(model_path, std::ios::binary);
            
            if (file.is_open()) {
                size_t size = model_data.size();
                file.write(reinterpret_cast<const char*>(&size), sizeof(size));
                file.write(reinterpret_cast<const char*>(model_data.data()), size * sizeof(double));
                log("ML model saved successfully");
            }
        } catch (const std::exception& e) {
            log("Error saving ML model: " + std::string(e.what()));
        }
    }

public:
    Features extract_features(const std::vector<double>& traffic_data) {
        Features features;
        
        if (traffic_data.empty()) {
            return features;
        }

        // Calculate basic statistics
        double sum = 0.0;
        double sum_sq = 0.0;
        features.min = traffic_data[0];
        features.max = traffic_data[0];

        for (double value : traffic_data) {
            sum += value;
            sum_sq += value * value;
            features.min = std::min(features.min, value);
            features.max = std::max(features.max, value);
        }

        features.mean = sum / traffic_data.size();
        
        // Calculate standard deviation
        double variance = (sum_sq / traffic_data.size()) - (features.mean * features.mean);
        features.std = std::sqrt(std::max(0.0, variance));

        // Calculate percentiles
        std::vector<double> sorted_data = traffic_data;
        std::sort(sorted_data.begin(), sorted_data.end());
        
        features.median = sorted_data[sorted_data.size() / 2];
        features.q25 = sorted_data[sorted_data.size() / 4];
        features.q75 = sorted_data[3 * sorted_data.size() / 4];

        // Calculate rate of change
        if (traffic_data.size() > 1) {
            double total_change = 0.0;
            for (size_t i = 1; i < traffic_data.size(); ++i) {
                total_change += traffic_data[i] - traffic_data[i - 1];
            }
            features.rate_of_change = total_change / (traffic_data.size() - 1);
        }

        return features;
    }

    std::pair<bool, double> analyze_traffic(double current_traffic) {
        std::lock_guard<std::mutex> lock(history_mutex);

        // Add current traffic to history
        TrafficData data;
        data.timestamp = std::chrono::system_clock::now();
        data.value = current_traffic;
        
        traffic_history.push_back(data);

        // Clean old data
        auto cutoff_time = std::chrono::system_clock::now() - 
                          std::chrono::seconds(config.analysis.time_window);
        
        traffic_history.erase(
            std::remove_if(traffic_history.begin(), traffic_history.end(),
                [cutoff_time](const TrafficData& d) {
                    return d.timestamp < cutoff_time;
                }),
            traffic_history.end()
        );

        // Extract traffic values
        std::vector<double> traffic_values;
        for (const auto& entry : traffic_history) {
            traffic_values.push_back(entry.value);
        }

        // Extract features
        Features features = extract_features(traffic_values);

        // Predict anomaly
        if (config.analysis.ml_enabled) {
            return predict_anomaly_ml(features);
        } else {
            return predict_anomaly_basic(features);
        }
    }

private:
    std::pair<bool, double> predict_anomaly_ml(const Features& features) {
        // Simulate ML prediction
        double score = calculate_anomaly_score(features);
        bool is_anomaly = score < -config.analysis.sensitivity;
        
        return {is_anomaly, std::abs(score)};
    }

    std::pair<bool, double> predict_anomaly_basic(const Features& features) {
        bool is_anomaly = (features.mean > config.thresholds.warning ||
                          features.max > config.thresholds.critical);
        
        double score = features.mean / config.thresholds.warning;
        
        return {is_anomaly, score};
    }

    double calculate_anomaly_score(const Features& features) {
        // Simulate isolation forest scoring
        double score = 0.0;
        
        // Combine features with weights
        score += features.mean * 0.3;
        score += features.std * 0.2;
        score += features.max * 0.2;
        score += features.rate_of_change * 0.3;
        
        // Normalize score
        score = (score - 50.0) / 25.0; // Assuming normal range
        
        return -score; // Negative for anomalies
    }

public:
    std::vector<Pattern> detect_patterns(const std::vector<TrafficData>& traffic_history) {
        std::vector<Pattern> patterns;

        if (traffic_history.size() < 10) {
            return patterns;
        }

        // Extract values
        std::vector<double> values;
        std::vector<std::chrono::system_clock::time_point> timestamps;
        
        for (const auto& entry : traffic_history) {
            values.push_back(entry.value);
            timestamps.push_back(entry.timestamp);
        }

        // Detect spikes
        detect_spikes(values, timestamps, patterns);

        // Detect trends
        detect_trends(values, timestamps, patterns);

        return patterns;
    }

private:
    void detect_spikes(const std::vector<double>& values, 
                      const std::vector<std::chrono::system_clock::time_point>& timestamps,
                      std::vector<Pattern>& patterns) {
        if (values.size() < 5) return;

        // Calculate rolling mean and std
        std::vector<double> rolling_mean, rolling_std;
        int window = 5;

        for (size_t i = window; i < values.size(); ++i) {
            double sum = 0.0, sum_sq = 0.0;
            for (int j = i - window; j < i; ++j) {
                sum += values[j];
                sum_sq += values[j] * values[j];
            }
            
            double mean = sum / window;
            double variance = (sum_sq / window) - (mean * mean);
            double std = std::sqrt(std::max(0.0, variance));
            
            rolling_mean.push_back(mean);
            rolling_std.push_back(std);
        }

        // Detect anomalies
        for (size_t i = 0; i < rolling_mean.size(); ++i) {
            double diff = std::abs(values[i + window] - rolling_mean[i]);
            if (diff > 2 * rolling_std[i]) {
                Pattern pattern;
                pattern.type = "spike";
                pattern.count = 1;
                pattern.max_value = values[i + window];
                pattern.timestamps.push_back(timestamps[i + window]);
                patterns.push_back(pattern);
            }
        }
    }

    void detect_trends(const std::vector<double>& values,
                      const std::vector<std::chrono::system_clock::time_point>& timestamps,
                      std::vector<Pattern>& patterns) {
        if (values.size() < 30) return;

        // Calculate trend using linear regression
        double sum_x = 0.0, sum_y = 0.0, sum_xy = 0.0, sum_x2 = 0.0;
        int n = values.size();

        for (int i = 0; i < n; ++i) {
            sum_x += i;
            sum_y += values[i];
            sum_xy += i * values[i];
            sum_x2 += i * i;
        }

        double slope = (n * sum_xy - sum_x * sum_y) / (n * sum_x2 - sum_x * sum_x);

        if (std::abs(slope) > config.thresholds.trend_threshold) {
            Pattern pattern;
            pattern.type = "trend";
            pattern.slope = slope;
            patterns.push_back(pattern);
        }
    }

public:
    std::string save_analysis(const Json::Value& analysis_data) {
        try {
            auto now = std::chrono::system_clock::now();
            auto time_t = std::chrono::system_clock::to_time_t(now);
            
            std::stringstream ss;
            ss << "analysis_" << std::put_time(std::localtime(&time_t), "%Y%m%d_%H%M%S") << ".json";
            
            std::string filename = ss.str();
            std::string filepath = rules_dir + "/" + filename;

            Json::StyledWriter writer;
            std::ofstream file(filepath);
            file << writer.write(analysis_data);
            
            log("Analysis saved to " + filepath);
            return filepath;
        } catch (const std::exception& e) {
            log("Error saving analysis: " + std::string(e.what()));
            return "";
        }
    }

    ThreatLevel get_threat_level(double anomaly_score) {
        if (anomaly_score > 0.8) return ThreatLevel::CRITICAL;
        if (anomaly_score > 0.6) return ThreatLevel::WARNING;
        if (anomaly_score > 0.4) return ThreatLevel::NOTICE;
        return ThreatLevel::NORMAL;
    }

    bool update_model(const std::vector<TrafficData>& new_data) {
        if (new_data.size() < 100) {
            return false;
        }

        try {
            std::vector<double> values;
            for (const auto& entry : new_data) {
                values.push_back(entry.value);
            }

            // Update model data (simplified)
            for (size_t i = 0; i < std::min(values.size(), model_data.size()); ++i) {
                model_data[i] = (model_data[i] + values[i]) / 2.0;
            }

            save_model();
            log("ML model updated successfully");
            return true;
        } catch (const std::exception& e) {
            log("Error updating model: " + std::string(e.what()));
            return false;
        }
    }

    // Configuration methods
    void set_sensitivity(double sensitivity) {
        config.analysis.sensitivity = sensitivity;
        ml_model.sensitivity = sensitivity;
    }

    double get_sensitivity() const {
        return config.analysis.sensitivity;
    }

    void set_ml_enabled(bool enabled) {
        config.analysis.ml_enabled = enabled;
    }

    bool is_ml_enabled() const {
        return config.analysis.ml_enabled;
    }

    // Statistics methods
    size_t get_history_size() const {
        std::lock_guard<std::mutex> lock(history_mutex);
        return traffic_history.size();
    }

    std::vector<TrafficData> get_recent_traffic(int count = 100) const {
        std::lock_guard<std::mutex> lock(history_mutex);
        
        std::vector<TrafficData> result;
        size_t start = (traffic_history.size() > count) ? 
                      traffic_history.size() - count : 0;
        
        for (size_t i = start; i < traffic_history.size(); ++i) {
            result.push_back(traffic_history[i]);
        }
        
        return result;
    }
};

// Main function for testing
int main() {
    TrafficAnalyzer analyzer;
    
    // Simulate traffic data
    std::random_device rd;
    std::mt19937 gen(rd());
    std::normal_distribution<double> normal_dist(50.0, 10.0);
    std::uniform_real_distribution<double> anomaly_dist(100.0, 200.0);
    
    for (int i = 0; i < 100; ++i) {
        double traffic_value;
        
        // Simulate occasional anomalies
        if (i % 20 == 0) {
            traffic_value = anomaly_dist(gen);
        } else {
            traffic_value = normal_dist(gen);
        }
        
        auto [is_anomaly, score] = analyzer.analyze_traffic(traffic_value);
        
        if (is_anomaly) {
            std::cout << "ANOMALY DETECTED! Score: " << score << std::endl;
        }
        
        std::this_thread::sleep_for(std::chrono::milliseconds(100));
    }
    
    return 0;
} 